url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/geo/genDataOfLineForHadoop.java

```diff
@@ -0,0 +1,151 @@
+/**
+ * 
+ */
+package utility.geo;
+
+/**
+ * @author qchen
+ *
+ */
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Scanner;
+import java.util.Set;
+
+import jdbm.PrimaryTreeMap;
+import jdbm.RecordManager;
+import jdbm.RecordManagerFactory;
+import utility.Compare.DataOfPoint;
+import utility.Compare.DataOfPointParser;
+import utility.Compare.buildBtreeOfPoints;
+
+/**
+ * remember first build buildBtreeOfPoints to get delaynay information
+ * @author qchen
+ * */
+public class genDataOfLineForHadoop {
+	public static boolean DEBUG = true;
+	public static String filename = "database/LinesData.NE.5HOP.tmp";
+	
+	public void loadData(int k) throws IOException{
+		RecordManager recmanOfPoints = RecordManagerFactory.createRecordManager(DataOfPointParser.filename);
+		PrimaryTreeMap<Long, byte[]> btOfPoints = recmanOfPoints.treeMap("treemap");
+		PrintWriter pw = new PrintWriter(new BufferedOutputStream(new FileOutputStream(new File(filename))));
+		Set<Long> keys = btOfPoints.keySet();
+		Queue<Long> qLongs = new LinkedList<Long>();
+		int timeToCommit = 0, constk = k;
+		for(Long key : keys){
+		//for(Long key = (long) 4000; key <= 5000; key ++){
+			pw.print(key);
+			k = constk;
+			DataOfPoint dataOfPoint = new DataOfPoint();
+			byte[] byteData = btOfPoints.find(key);
+			dataOfPoint.loadFromBytes(byteData);
+			HashSet<Long> cnt = new HashSet<Long>();
+			cnt.add(key);
+			for(Long id : dataOfPoint.delaunayIds){
+				qLongs.offer(id);
+				cnt.add(id);
+				k ++;
+			}
+			while(!qLongs.isEmpty() && k > 0 && constk >= 0){
+				Long id = qLongs.poll();
+				byteData = btOfPoints.find(id);
+				DataOfPoint tmpDataOfPoint = new DataOfPoint();
+				tmpDataOfPoint.loadFromBytes(byteData);
+				//ids.add(id);
+				//lines.add(new Line(dataOfPoint.p.doublePoint(), tmpDataOfPoint.p.doublePoint()));
+				pw.print(" " + id + " " + dataOfPoint.p.doublePoint().x + " " + dataOfPoint.p.doublePoint().y + " " 
+						+ tmpDataOfPoint.p.doublePoint().x + " " + tmpDataOfPoint.p.doublePoint().y);
+				for(Long tmpLong : tmpDataOfPoint.delaunayIds){
+					if(cnt.contains(tmpLong) || qLongs.size() >= k)continue;
+					qLongs.offer(tmpLong);
+					cnt.add(tmpLong);
+				}
+				k --;
+			}
+			pw.println();
+			//byte[] testdata = newDataOfLine.writeToBytes();
+			timeToCommit ++;
+			qLongs.clear();
+		}
+		recmanOfPoints.close();
+		pw.close();
+	}
+	
+	public void loadDataByhop(int k) throws IOException{
+		RecordManager recmanOfPoints = RecordManagerFactory.createRecordManager(DataOfPointParser.filename);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *recmanOfPoints*. The resource is referenced at line 87. The resource is closed at line 127. There are other execution paths that do not contain closure statements, for example, when DataOfPoint.loadFromBytes throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).